### PR TITLE
Sampling, batch_all, non-zero, optimizer-flag

### PR DESCRIPTION
**Pull request by lucasb-eyer**
So this is a branch I had lying around for a while which adds a lot of things. I'm not merging it yet, as I still want to add these changes to the README and test it thoroughly, but here it is for others to try out.

It implements the trick I mentioned in #4 as well as some more variants we cover in the paper. If you give this a try, please give feedback here.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -294,16 +303,24 @@ def main():
...
         dists, pids, args.margin, batch_precision_at_k=args.batch_k-1)
 
-    # Count the number of active entries, and compute the total batch loss.
-    num_active = tf.reduce_sum(tf.cast(tf.greater(losses, 1e-5), tf.float32))
-    loss_mean = tf.reduce_mean(losses)
+    # Count how many entries in the batch are (possibly approximately) non-zero.
+    if args.loss_ignore_zero is True:
+        nnz = tf.count_nonzero(losses, dtype=tf.float32)
+    else:
+        nnz = tf.reduce_sum(tf.to_float(tf.greater(losses, args.loss_ignore_zero or 1e-5)))
```
Comment (maxisme) [inline train.py]: Is the point of this supposed to be just for logging?

Comment (lucasb-eyer) [reply train.py]: No, actually. It's type-magic and can be a little obscure, hence why I still need to write documentation in the README :)

The `else` case happens when `loss_ignore_zero` is given an additional float argument, so one can call it as `--loss_ignore_zero 1e-3` for example, in order to consider anything below 1e-3 to be counted as zero.

File: train.py [hunk 2]
```diff
@@ -102,9 +102,13 @@
     help='Which metric to use for the distance between embeddings.')
 
 parser.add_argument(
-    '--loss', default='batch_hard', choices=loss.LOSS_CHOICES.keys(),
+    '--loss', default='batch_hard', choices=loss.LOSS_CHOICES,
     help='Enable the super-mega-advanced top-secret sampling stabilizer.')
 
+parser.add_argument(
+    '--loss_ignore_zero', default=False, const=True, nargs='?', type=common.positive_float,
```
Comment (maxisme) [inline train.py]: Ohh misread this to mean it can only be true or false. I am going to start playing with this then.

File: train.py (hunk 1 above)
Comment (maxisme) [reply train.py]: Just trying to get my head around this. Why would making loss' that aren't zero, zero help anything?

Comment (lucasb-eyer) [reply train.py]: Read our paper, we explain them in there :) But really it's not a good time investment to play with that parameter.
